### PR TITLE
Fogl 4591: Multiple readings dbs/tables configurable parameters

### PR DESCRIPTION
**Pull request by sstefano69**
*Note*:
* The parameters will be changed only if they are compatible in relations to the data loaded/used by the user
* new readings tables naming
* configurable using API or GUI
* after the configuration a restart of fledge is needed

Sampli API call:
```

    curl -s -X PUT http://localhost:8081/fledge/category/sqlite/nDbPreallocate                   -d '{ "value" : "4" }'
    curl -s -X PUT http://localhost:8081/fledge/category/sqlite/nReadingsPerDb                   -d '{ "value" : "20" }'


    curl -s -X PUT http://localhost:8081/fledge/category/sqlite/nDbLeftFreeBeforeAllocate       -d '{ "value" : "1" }'
    curl -s -X PUT http://localhost:8081/fledge/category/sqlite/nDbToAllocate                   -d '{ "value" : "1" }'

```

*New dbs/tables naming* - 3 tables per db:

image

image

Db1:
image

image

Db2:
image

image

### CODE REVIEW

File: C/plugins/storage/sqlite/plugin.cpp [hunk 1]
```diff
@@ -40,8 +41,37 @@ const char *default_config = QUOTE({
...
 			"displayName" : "Pool Size",
 			"order" : "1"
-			}
-		});
+		},
+		"nReadingsPerDb" : {
+			"description" : "Number of readings tables per  database",
+			"type" : "integer",
+			"default" : "15",
+			"displayName" : "N Readings per database",
```
Comment (MarkRiddoch) [inline C/plugins/storage/sqlite/plugin.cpp:49]: We need to keep display names short but meaningful. Either change "N" to be "No." or drop it completely "Readings per database".

File: C/plugins/storage/sqlite/plugin.cpp [hunk 2]
```diff
@@ -40,8 +41,37 @@ const char *default_config = QUOTE({
...
+			"order" : "2"
+		},
+		"nDbPreallocate" : {
+			"description" : "Number of databases to allocate in advance, NOTE: SQLite has a maximum number of attachable databases by default at 10",
```
Comment (MarkRiddoch) [inline C/plugins/storage/sqlite/plugin.cpp:53]: "Number of databases to allocate in advance. NOTE: SQLite has a default maximum of 10 attachable databases"

File: C/plugins/storage/sqlite/plugin.cpp [hunk 3]
```diff
@@ -40,8 +41,37 @@ const char *default_config = QUOTE({
...
+			"type" : "integer",
+			"default" : "3",
+			"displayName" : "N databases to allocate in advance",
```
Comment (MarkRiddoch) [inline C/plugins/storage/sqlite/plugin.cpp:56]: "No. preallocated databases"

File: C/plugins/storage/sqlite/plugin.cpp [hunk 4]
```diff
@@ -40,8 +41,37 @@ const char *default_config = QUOTE({
...
+			"order" : "3"
+		},
+		"nDbLeftFreeBeforeAllocate" : {
+			"description" : "Number of databases left free before a new allocation is executed",
```
Comment (MarkRiddoch) [inline C/plugins/storage/sqlite/plugin.cpp:60]: "Allocate new databases when the number of free databases drops below this value"

File: C/plugins/storage/sqlite/plugin.cpp [hunk 5]
```diff
@@ -40,8 +41,37 @@ const char *default_config = QUOTE({
...
+			"type" : "integer",
+			"default" : "1",
+			"displayName" : "N free databases before allocation",
```
Comment (MarkRiddoch) [inline C/plugins/storage/sqlite/plugin.cpp:63]: "Database allocation threshold"

File: C/plugins/storage/sqlite/plugin.cpp [hunk 6]
```diff
@@ -40,8 +41,37 @@ const char *default_config = QUOTE({
...
+			"order" : "4"
+		},
+		"nDbToAllocate" : {
+			"description" : "Number of databases to allocate each time",
+			"type" : "integer",
+			"default" : "2",
+			"displayName" : "N databases to allocate",
```
Comment (MarkRiddoch) [inline C/plugins/storage/sqlite/plugin.cpp:70]: "Database allocation size"

File: C/plugins/storage/sqlite/plugin.cpp [hunk 7]
```diff
@@ -40,8 +41,37 @@ const char *default_config = QUOTE({
...
+		},
+		"nDbLeftFreeBeforeAllocate" : {
+			"description" : "Number of databases left free before a new allocation is executed",
+			"type" : "integer",
+			"default" : "1",
+			"displayName" : "N free databases before allocation",
+			"order" : "4"
+		},
+		"nDbToAllocate" : {
+			"description" : "Number of databases to allocate each time",
```
Comment (MarkRiddoch) [inline C/plugins/storage/sqlite/plugin.cpp:67]: "The number of databases to create whenever the number of available databases drops below the allocation threshold"

File: C/plugins/storage/sqlite/plugin.cpp [hunk 8]
```diff
@@ -40,8 +41,37 @@ const char *default_config = QUOTE({
 			"default" : "5",
 			"displayName" : "Pool Size",
 			"order" : "1"
-			}
-		});
+		},
+		"nReadingsPerDb" : {
+			"description" : "Number of readings tables per  database",
```
Comment (MarkRiddoch) [inline C/plugins/storage/sqlite/plugin.cpp:46]: "The number of readings tables in each database that is created"

**Review by MarkRiddoch — CHANGES_REQUESTED** [review]
Some minor rewording

File: C/plugins/storage/sqlite/plugin.cpp (hunk 2 above)
Comment (sstefano69) [reply C/plugins/storage/sqlite/plugin.cpp]: done

File: C/plugins/storage/sqlite/plugin.cpp (hunk 3 above)
Comment (sstefano69) [reply C/plugins/storage/sqlite/plugin.cpp]: done

File: C/plugins/storage/sqlite/plugin.cpp (hunk 4 above)
Comment (sstefano69) [reply C/plugins/storage/sqlite/plugin.cpp]: done

File: C/plugins/storage/sqlite/plugin.cpp (hunk 5 above)
Comment (sstefano69) [reply C/plugins/storage/sqlite/plugin.cpp]: done

File: C/plugins/storage/sqlite/plugin.cpp (hunk 6 above)
Comment (sstefano69) [reply C/plugins/storage/sqlite/plugin.cpp]: done

File: C/plugins/storage/sqlite/plugin.cpp (hunk 7 above)
Comment (sstefano69) [reply C/plugins/storage/sqlite/plugin.cpp]: done

File: C/plugins/storage/sqlite/plugin.cpp (hunk 8 above)
Comment (sstefano69) [reply C/plugins/storage/sqlite/plugin.cpp]: done

**Review by pintomax — APPROVED** [review]

**Review by YashTatkondawar — APPROVED** [review]
# Other than FOGL-4778 everything looks good to me.

**Pull request merged**